Basic framework for generic breakable items.

Hinaichigo opened the pull request:
This is a basic framework for breakable items. You can set any item to be breakable in different contexts, including a mob attacking it. The variables it uses, with some example values, are:

Breakability:
`breakable_flags = BREAKABLE_HIT`
Flags for situations the item can take damage or break in. See breakable_defines.dm for the full list
`health = 15`
Structural integrity of the item. At 0, the item breaks.
`maxHealth = 15`
`damage_armor = 5`
Attacks of this much damage or below will glance off
`damage_resist = 5`
Attacks stronger than damage_armor will have their damage reduced by this much
`breakable_exclude = list(/obj/item/tool/weldingtool)`
List of specific items that will bypass breaking the item on harm intent, so that unique, item-specific harm intent interactions can still occur. 

Fragmentation:
`breakable_fragments = list(/obj/item/weapon/shard, /obj/item/weapon/shard/plasma)`
List of item types that will be produced when the item is broken apart.
`fragment_amounts = list(2,1)`
List of the amounts of each fragment type that will be produced upon the item breaking, defaults to 1 each.

Text:
`take_hit_text = "cracking" `
Visible message when the item is hit and damaged but not fully broken. eg. "chipping" becomes "..., chipping it!"
`take_hit_text2 = "cracks" `
Another visible message when the item is hit and damaged but not fully broken. eg. "chips" becomes "The ... chips!"
`damaged_text = "It looks cracked."`
Addendum to the examine description when the item is damaged
`breaks_text = "shatters"`
Visible message when the items breaks. eg. "breaks apart".
`glances_text = "bounces"`
Visible message when an attack glances off the item. eg. "glances" yields "... but it glances off!"

Sounds:
`breaks_sound = 'sound/effects/Glassbr3.ogg'`
Sound when the item breaks.
`glanced_sound`
Sound when the item receives a glancing attack.
`damaged_sound`
Sound when the item recieves an attack that damages but doesn't break it.



This was initially so that grues could smash desk lamps, but as a general framework could be useful to make various items smashable without having to add custom behavior each time.
So far it's set to work for humanoids kicking or biting the item, simple mobs attacking the item, attacking the item with a weapon, the item ballistically colliding with something, and the item receiving recoil damage when used as a melee weapon. Eschews empty-handed hitting of items for humans, because that conflicts with picking up in many contexts. Plus humans usually would have access to kicking anyway. If a specific punch-interaction is added like with kicking and biting, it could work though.

I'd like to ask if anyone knows of some other case or cases that I'm missing, or if things can somehow be generalized more, and also for feedback in general.

There are some testing items included; I can take them out if it's otherwise ready to merge.

Features:
- Items breaking when attacked with another item.
- Items breaking when hit by a simple mob.
- Items breaking when shot by a projectile.
- Items breaking when thrown against something.
- Items breaking when used as a weapon.
- Items breaking when kicked.
- Items breaking when bitten.
- Item armor and resistance against breaking.
- Items breaking into fragments when broken, transferring bloodstains, suit fibers, and fingerprints.
- Fragments being propelled if appropriate when the item breaks upon being kicked or shot by a projectile.
- Specifiable visible message text and sounds when an item breaks, is damaged without fully breaking, or receives a glancing attack not strong enough to damage it.
- Contents of a container item spilling when it breaks, being propelled if appropriate.
- Reagents inside an item spilling out onto what it broke upon.

Areas for future expansion:

- Integrate with existing implementations of item health.
- Generalize to /obj.
- Different types of resistances and responses to different types of damage. (eg. fork cracking when shot with a bullet but melting when shot with a laser)
- Integrate existing shrapnel system into fragments.
- Transferring over-damage into fragments for further breaking.
- Designating a specific fragment to remain in the item's inventory slot when the item breaks. (eg. baseball bat splintering apart but remaining in the hand)
- Getting hurt when holding an item as it breaks.
- Getting hurt when biting or kicking an item in certain situations.
- Interaction with explosions, extreme temperatures, shuttle crushing, and other forms of damage.
- Transferring damage to container contents without damaging the container where applicable. (eg. hitting a cloth bag full of beakers with a crowbar)
- Damage considerations for the strength of the weapon wielder.
- Recoil damage considerations for the hardness of the target being attacked.
- Icon overlays for damaged items.
- Implement a generic fragment item with a procedural icon.
- Make breakability into a component.
- Repairing damaged items.

[content]
:cl:
 * rscadd: Designated items can be smashed and broken.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+//use kicking code
+//generalize out of devices
+
+/obj/item/device
```
```suggestion
/obj/item
```
/device is a silly subtype. If you're going to do this, do it on the base `/obj/item` type or even `/obj`.
In fact, ideally this would be a component.

Hinaichigo replied:
Could you please let me know what you mean by component here?

DamianX replied:
A `/datum/component` that can be `add_component`ed to anything. It's really not easy to work with so an inheritance-based approach is fine too, if you don't want to do it that way.

Hinaichigo replied:
I generalized it to `/obj/items` for now. I'll consider the ramifications of smashing other types of `/obj`.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+/obj/item/device
+
+	//Destructability parameters:
+	var/breakable = 0 //1: breakable (by smashing it using the following flags
```
Use TRUE and FALSE whenever the expected values are only 1 and 0.
Also, isn't this redundant because `breakable_flags = 0` exists?

Hinaichigo replied:
I was thinking a situation could arise where someone would want to toggle an item's breakability off without setting the flags to 0, maybe if they wanted to only disable the breakability temporarily so another harm-intent action could be performed, and then turn it back on later.

Hinaichigo replied:
I think I'll remove `breakable` and instead add a list of excluded items that bypass the breaking so that specific harm intent interactions can still be performed on an item while the item still remains generally breakable.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+	var/breakable = 0 //1: breakable (by smashing it using the following flags
+	var/breakable_flags = 0 /*possible flags include BREAKABLE_ALL | BREAKABLE_HIT | BREAKABLE_HIT_EMPTY | BREAKABLE_HIT_WEAPON | BREAKABLE_THROW
+							BREAKABLE_HIT encompasses both BREAKABLE_HIT_EMPTY and BREAKABLE_HIT_WEAPON */
+	var/struct_integ = 15 //structural integrity of the item, akin to HP.
```
Something like `var/health` would convey the same meaning and be shorter.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+	var/damaged_text = "" //Addendum to the description when it's damaged eg. damaged_text of "It is dented." Empty string "" will skip this addendum.
+	var/breaks_text = "" //Visible message when the items breaks. eg. "breaks apart" Empty string skips this.
+	var/breaks_sound = "" //path to audible sound when the item breaks. Empty string skips this.
+	var/list/breakable_fragments = list() //List of objects that will be produced when the item is broken apart. eg. /obj/weapon/item/shard
```
Creating an empty list for each item is not the end of the world but also not great for memory usage. You should leave it uninitialized (which defaults to null) and only create it when it's used. The code using this will have to be made aware of this.

Additionally, I don't know if a simple list of things to spawn will be flexible enough for a system like this. I think a lot of implementations will want custom behavior to transfer fingerprints and whatnot from the item being destroyed.

Hinaichigo replied:
I'll take a look at fingerprint transferring.
Any specific custom behavior could be added in `on_broken()` for that item, but aside from fingerprints and bloodstains, is there anything else that should be handled by the default `drop_fragments()`?

Hinaichigo replied:
For now it transfers fingerprints and suit fibers, as well as blood information and generates a blood overlay if the item was bloody itself. So smashing a blood-stained beaker would result in blood-stained shards.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+	var/damage_resist = 5 //attacks stronger than damage_armor will have their damage reduced by this much
+	var/damaged_text = "" //Addendum to the description when it's damaged eg. damaged_text of "It is dented." Empty string "" will skip this addendum.
+	var/breaks_text = "" //Visible message when the items breaks. eg. "breaks apart" Empty string skips this.
+	var/breaks_sound = "" //path to audible sound when the item breaks. Empty string skips this.
```
```suggestion
	var/breaks_sound // path to audible sound when the item breaks. null skips this.
```

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+/obj/item/device/proc/on_broken() //Called right before an object breaks
+	if(breaks_text!="")
+		visible_message("<span class='notice'>\The [src] [breaks_text]!</span>")
+	if(breaks_sound!="")
```
```suggestion
	if(!breaks_sound)
```

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+
+//Aliens
+/obj/item/device/attack_alien(mob/living/carbon/alien/humanoid/user)
+	if(user.a_intent == I_HURT && breakable && breakable_flags | BREAKABLE_HIT_EMPTY)
```
The correct syntax is `value & FLAG_TO_CHECK`. As it is right now, `breakable_flags | BREAKABLE_HIT_EMPTY` would always be true.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+/obj/item/device/attack_alien(mob/living/carbon/alien/humanoid/user)
+	if(user.a_intent == I_HURT && breakable && breakable_flags | BREAKABLE_HIT_EMPTY)
+		user.do_attack_animation(src, user)
+		user.delayNextAttack(10)
```
```suggestion
		user.delayNextAttack(1 SECONDS)
```

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,129 @@
+	if(user.a_intent == I_HURT && breakable && breakable_flags | BREAKABLE_HIT_EMPTY)
+		user.do_attack_animation(src, user)
+		user.delayNextAttack(10)
+		receive_damage(user.get_unarmed_damage()) ? user.visible_message("<span class='warning'>\The [user] [pick("slashes","claws")] \the [src]!</span>","<span class='notice'>You [pick("slash","claw")] \the [src]!</span>") : user.visible_message("<span class='warning'>\The [user] [pick("slash","claw")] \the [src], but it [pick("bounces","gleams","glances")] off!</span>","<span class='notice'>You hit \ ...
```
Quite unwiedly. I would turn this into an if-statement

DamianX commented:
I have considered implementing something like this a couple of times but never went through with it. I think I remember TG's implementation was somewhat decent, you may want to take inspiration from that. (Note that TG's license is incompatible with /vg/'s).

Hinaichigo commented:
Upon changing `struct_integ` to `health`, it seems that `health` was already defined for items, initialized to `null`. I could use the existing var but I wonder if it could break anything if that's actually used elsewhere. What do you think @DamianX?

DamianX commented:
Rename it, then. `obj_health` maybe? You can look at the types that currently define a `health` var to see what kind of things they do with it, and see if they'll fit into the new system

Hinaichigo commented:
> Rename it, then. `obj_health` maybe? You can look at the types that currently define a `health` var to see what kind of things they do with it, and see if they'll fit into the new system

Do you know a good way I could search the code for that? "health" is such a commonly used phrase that it's tricky to filter out item-specific cases.

DamianX commented:
A crude way to do it is to define it on /obj or /obj/item and try to compile and look at the error.
Or you can use the visual studio code extension for DM and hit Ctrl-T and type "health".

Hinaichigo commented:
> A crude way to do it is to define it on /obj or /obj/item and try to compile and look at the error. Or you can use the visual studio code extension for DM and hit Ctrl-T and type "health".

Thanks.

Hinaichigo commented:
There's a bug where trying to pick up an item on harm intent also counts as you punching it.

Hinaichigo commented:
I think it might be actually better to leave out the open-handed punching of objects for humans, because with kicking they still usually have a way to smash objects unarmed, and it won't conflict with them trying to pick up an item in the heat of the moment: say trying to quickly pick up a weapon on harm intent and then attack someone; a player wouldn't want to accidentally punch the item and then try to hastily change to help intent as their foe closes in on them. In other words, harm intent shouldn't necessarily exclude picking up items, and with the ability to kick and break items, they already usually have a means to smash items without a weapon.

Also now if kicking breaks the item apart, the fragments will be propelled away. I'm thinking about integrating the existing shrapnel system into the fragments as well.

Hinaichigo commented:
@DamianX If you wouldn't mind taking a look at `obj/item/kick_act()` in `items.dm` in particular, I'd appreciate it. I sort of want to make this more general and automatic so that you can do things like use a baseball bat to smash something out of the air and then have the fragments fly off properly, without having to explicitly define it each time, but I'm still thinking of a way to do that.

Hinaichigo commented:
Okay, implemented all the basic features I was thinking of.

Hinaichigo commented:
@DamianX If you could please take a look at this it'd be very much appreciated.

DamianX commented on __DEFINES/setup.dm:
```diff
@@ -1802,3 +1802,22 @@ var/list/weekend_days = list("Friday", "Saturday", "Sunday")
 #define SPORTINESS_SPORTS_DRINK 5
+
+//Breakable items
+#define BREAKABLE_ALL 15			//Item can break by being hit, or when used to hit something as a melee or a thrown weapon.
+#define BREAKABLE_HIT	3			//Item can break by being hit with either an unarmed attack or a weapon.
+#define BREAKABLE_UNARMED 1			//Item can break by being hit with an unarmed attack.
+#define BREAKABLE_WEAPON 2			//Item can break by being hit with a weapon. This includes projectiles.
+#define BREAKABLE_AS_ALL 12			//Item can break when it's used as a melee weapon, or when it ballistically collides with something.
```
You can organize these macros better:
- Move them to their own file under `__DEFINES/`
- `BREAKABLE_ALL` could just be defined as `#define BREAKABLE_ALL ALL`, I think. So that adding and removing flags in the future doesn't require updating `BREAKABLE_ALL` as well.
- The flags could be ordered from smallest to biggest
- Compound flags such as `BREAKABLE_AS_ALL` could be written by combining previous flags (and therefore necessarily declared after them) like: `#define BREAKABLE_AS_ALL (BREAKABLE_WEAPON | BREAKABLE_AS_THROWN)` and so on.
- The values could be expressed in the form of bitshifts, so that you get a nice list of ordered contiguous integers like this:
```
#define BREAKABLE_UNARMED	(1<<0)
#define BREAKABLE_WEAPON	(1<<1)
```
and so on.

Hinaichigo replied:
> * `BREAKABLE_ALL` could just be defined as `#define BREAKABLE_ALL ALL`, I think. So that adding and removing flags in the future doesn't require updating `BREAKABLE_ALL` as well.

Where does the ALL come from, what does it mean?

DamianX replied:
It's defined here https://github.com/vgstation-coders/vgstation13/blob/5c9b1517d08ac851ab37b0f6b13127c230149c26/__DEFINES/setup.dm#L212
It means every bitflag is set.

Hinaichigo replied:
Is that equivalent to ...11111111? I'm not sure that it'd work in this case because there is a flag BREAKABLE_NOMOB that overrides other flags and is excluded from BREAKABLE_ALL.

DamianX replied:
Should that be a thing? Why do it like that instead of a flag like BREAKABLE_MOB that works like the others?

Hinaichigo replied:
That makes sense, I changed it.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,263 @@
+
+/obj/item
+	//Breakability:
+	var/breakable_flags /*Possible flags include BREAKABLE_ALL, BREAKABLE_HIT, BREAKABLE_UNARMED, BREAKABLE_WEAPON, BREAKABLE_AS_ALL, BREAKABLE_AS_THROWN, BREAKABLE_AS_MELEE, and BREAKABLE_NOMOB.
```
I would just have the comment point to the file that contains the macros. No one is going to keep this comment updated.

Hinaichigo replied:
Done.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,263 @@
+/obj/item/proc/drop_contents(var/atom/target, var/range, var/speed, var/override, var/fly_speed) //Drop the contents of the item and propel them if the item itself received a propulsive blow.
+	if(contents.len)
+		for(var/obj/item/thiscontent in contents)
+			thiscontent.loc = src.loc
```
Use `forceMove` instead of setting loc directly, unless you have a good reason.

DamianX commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,263 @@
+	if(!isnull(reagents))
+		if(!isnull(hit_atom) && hit_atom != get_turf(src)) //If it hit something other than the floor, spill it onto that.
+			reagents.reaction(hit_atom, TOUCH)
+		reagents.reaction(get_turf(src), TOUCH) //Then spill it onto the floor.
```
Do you need an `else` here? Otherwise the reagents would spill onto a movable first, and then onto the floor. Is this intended? Does it even work? Aren't the reagents consumed by the first `reaction`?

Hinaichigo replied:
The idea is that it first attempts to react anything in the item onto the thing it hit. Then whatever didn't get consumed in the that reaction spills onto the floor. It seems to work from my testing. A beaker full of blood hits a cyborg and the blood then leaves a stain on the floor. A polyacid beaker hits a person and splatters them with acid, burning them.

DamianX commented on code/_onclick/item_attack.dm:
```diff
@@ -236,11 +236,20 @@
+	//Break the item if applicable.
+	if(power && (I.breakable_flags & BREAKABLE_AS_MELEE) && !(I.breakable_flags & BREAKABLE_NOMOB))
+		take_damage(min(power, BREAKARMOR_MEDIUM), FALSE) //Cap recoil damage at BREAKARMOR_MEDIUM to avoid a powerful weapon also needing really strong armor to avoid breaking apart when used. Be verbose about the item being damaged if applicable.
+		break_item(null, null, null, null, null, M) //Break the item and spill any reagents onto the target.
```
This is quite hard to read. You can use named arguments like so:
```suggestion
		break_item(hit_atom = M) //Break the item and spill any reagents onto the target.
```
The rest of the arguments will be null.

Hinaichigo replied:
That's good to know.

DamianX reviewed (commented):
Haven't read all of it yet

DamianX commented:
I'd like someone else to take a look at this @Exxion @Kurfursten @jknpj

Hinaichigo commented:
I changed the variable `health_item` to use the existing `health` variable, and changed `item_health_max` to `maxHealth`.
This way it will be more easy to generalize breakability to other types of `obj`and perhaps other atoms down the line, For instance, closets already have a `health` variable that allows them to be shot by projectiles. Since items that already use `health` won't have `breakable_flags` set, there shouldn't be any conflicts with existing systems.

Hinaichigo commented:
Is this still being looked over?

Exxion commented:
I keep forgetting to take a look. I'll try to do it tomorrow.

Exxion commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,260 @@
+		playsound(src, damaged_sound, 50, 1)
+
+/obj/item/proc/drop_fragments(var/atom/target, var/range, var/speed, var/override, var/fly_speed) //Separate proc in case special stuff happens with a given item's fragments. Parameters are for throwing the fragments.
+	if(breakable_fragments.len)
```
```suggestion
	if(breakable_fragments?.len)
```
Otherwise it will runtime on objects for which the var is null.

Exxion commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,260 @@
+	if(isobserver(user) || !Adjacent(user) || user.is_in_modules(src))
+		return
+
+	if(user.a_intent == I_HURT && breakable_flags & BREAKABLE_WEAPON && loc != user) //Smash items on the ground, but not in your inventory.
```
Can't break an item directly in your mob, but you can if it's in a container?

Exxion commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,260 @@
+	else
+		return "!" //Don't say "cracking it" if it breaks because on_broken() will subsequently say "The item shatters!"
+
+/obj/item/proc/break_item(var/atom/target, var/range, var/speed, var/override , var/fly_speed, var/hit_atom) //Breaks the item if its health is 0 or below. Passes throw-related parameters to on_broken() to allow for an object's fragments to be thrown upon breaking.
```
This proc is rather poorly-named. It checks if the item should break and breaks it if so, while the name would suggest that it just breaks it.

Exxion commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,260 @@
+			reagents.reaction(hit_atom, TOUCH)
+		reagents.reaction(get_turf(src), TOUCH) //Then spill it onto the floor.
+
+/obj/item/proc/take_damage(var/incoming_damage, var/mute = TRUE)
```
It's very strange that `take_damage()` would not also handle breaking the item if its health drops to/below zero. I see the places where there's code between the `take_damage()` and `break_item()` but I can't imagine there's not a better way to handle that.

Hinaichigo replied:
One reason is that the text generation for various types of attacks needs to know in advance what happened to the item when it was attacked. If it happens out of order, then the message about attacking the item will occur after the item breaks, leading to things like:
> You attack the breaker with the crowbar, cracking it!
> The beaker shatters!
> You attack the beaker with the crowbar!

To say the attack text within `take_damage()` or `break_item()` would require each new type of attack or damage source to be accounted for within those procs, or a flag set on trying to break the item or not in each situation. Whichever way, its somewhat complicated but keeping the functionality of the procs simple and separate is why this way seems to me to be optimal. But maybe there's a good solution?

Exxion replied:
The cases where there's actual behavior between the two procs are pretty unavoidable. They do have to be called separately for those. The cases with just messages *could* be handled inside the procs, which would be simple and clean if we had real format strings, but we don't. If it weren't for the cases that required them to be called separately, I'd say handle them inside anyway, but since they have to be separated sometimes anyway it's fine.
However, none of this stops you from calling `break_item()` inside `take_damage()` by default and having an arg that tells it not to, or wrapping them both in a different proc to be used when possible, or something. Remember that other people work on the codebase, and having to separately call two procs in all cases just because they have to be separated in a few isn't very intuitive.

Exxion commented on code/game/objects/items/breakable.dm:
```diff
@@ -0,0 +1,260 @@
+			take_damage(thisdmg/2)
+		else
+			take_damage(thisdmg, FALSE) //Be verbose about the item taking damage.
+	break_item(null,null,null,null,null,hit_atom)
```
Probably use a named arg instead

Hinaichigo replied:
> Also, the code style is very inconsistent and bad, but since we _still_ don't have an official style guide, I don't get to officially complain about it.

Just for my own edification, what kind of things would be good practice in terms of code style?

Exxion commented on code/game/objects/items/weapons/storage/storage.dm:
```diff
@@ -364,6 +364,12 @@
+
+	//Allow smashing of storage items on harm intent without also putting the weapon into the container.
+	var/mob/user = usr
+	if (user.a_intent == I_HURT && breakable_flags & BREAKABLE_WEAPON && loc != user)
```
Runtimes if `usr` does not exist. Why not just put it under the `if(usr)` right below?

Exxion commented:
It's strange to first apply this to items then later extend it to objs rather than first set up the system for objs then add specific behavior for items, but this is what's here.

Also, the code style is very inconsistent and bad, but since we *still* don't have an official style guide, I don't get to officially complain about it.

Hinaichigo commented:
> Also, the code style is very inconsistent and bad, but since we still don't have an official style guide, I don't get to officially complain about it.

For my own edification, what kind of things would be good practice in terms of code style?

Exxion commented:
My biggest gripe is the lack of spaces around some binary operators. So `a=b` instead of `a = b` and the like. This is bad in particular with `a=!b` instead of `a = !b`. I'm also not fond of the intermittent lack of spaces after commas and some other stuff, but that's all really just nitpicks. It's also that it's not consistent. But again, I'm not going to hold this up over style issues since we don't write this stuff down anywhere and there are lots of inconsistencies already present.

Hinaichigo commented:
I implemented the changes you recommended. Also, I put all the propulsion parameters (from `throw_at()`) into a datum, but I'm not sure if that's the best practice. I think it could make things more modular down the line though. Please let me know what you think.

DamianX commented on code/game/objects/items/weapons/storage/storage.dm:
```diff
@@ -365,12 +365,15 @@
 	if(usr) //WHYYYYY
+
+		//Allow smashing of storage items on harm intent without also putting the weapon into the container.
+		var/mob/user = usr
```
Is there a point to this? `usr` is already a mob

Hinaichigo replied:
I was advised in another PR review to avoid directly referencing `usr` wherever possible. But maybe it would be better to avoid an unnecessary variable declaration?

DamianX replied:
You need to avoid `usr` if you have the option of getting a reference to the mob you want as a parameter to your proc.
Rewriting `handle_item_insertion` and calls to it to have a `mob/user` parameter would be good, but out of scope.
Doing this cast would just add a line for no benefit.
The issue with `usr` is that it's a reference to the mob that started the current proc chain, so if *somehow* a player made another player insert something into a storage object, `usr` would not refer to the player who is actually performing the insertion. This has caused bugs in the past.

DamianX commented:
I guess we could give this a shot...?

ValkyrieSkies commented:
I can't comment on the code implementation, but conceptually this is a very good development - far too many things in the game are arbitrarily invulnerable.
